uat/testing-features/src/main/resources/greengrass/components/recipes/client_java_sdk.yaml

```diff
@@ -23,4 +23,4 @@ Manifests:
           Execute: ALL
     Lifecycle:
       Run: |
-        java -jar {artifacts:path}/aws.greengrass.client.Mqtt5JavaSdkClient.jar {configuration:/agentId} {configuration:/controlAddress} "{configuration:/controlPort}"
+        java -jar {artifacts:path}/aws.greengrass.client.Mqtt5JavaSdkClient.jar "{configuration:/agentId}" "{configuration:/controlPort}" {configuration:/controlAddresses}
```

Comment:
I think you'll want to JSON encode the control address array and pass it as stringified JSON, this will make parsing on the other end simpler and more extensible.

Reply:
IP addresses are passed as shell style program arguments.
Double quote prevents one but with spaces argument like "first argument" to be accepted as two separated arguments by the program. At the same time list of IP passed without quotas because passed as separated arguments. 
JSON useless here.

{configuration:/controlPort} is just reference to Greengrass component configuration value.
From scenario side these are settings of deployment passed as JSON doc.

Reply:
If you need to add any other arguments, then that becomes difficult since you are trying to handle the last _n_ items as IP address information.

Using named environment variables and/or JSON to pass the configuration would be less errorprone and require less manual work to parse the inputs.

Reply:
That is usual impediments for arguments of the CLI programs.
It is solved by using option signs like "-a IP -a IP -a IP"  or "--address IP --address IP" or  end-options-mark like "-- IP IP IP IP ".
But that structure of CLI arguments should be the same over all MQTT clients. When using JSON we need to add a JSON library to each client. For example C/C++ completely absent global packages repo like maven central that means with each new dependency you more and more limit your software to be build and run only on selected versions of OSes. Has no sence to inject JSON library only to pass list of string to the program.

Windows missed the default library to parse command line short or long options, that was a reason why arguments were extremely simplified.

Reply:
Using environment variables would also simplify this.

`CLIENT_ID=client id`, `IP_ADDRESSES="<list of ip space separated>"`, ... other config options.

Reply:
Environment variables are the implicit dependency, arguments - explicit.
In that case we need explicit dependency.
Complexity of passing value via environment is the same as aruments.
No pros.

Reply:
shell already did parsing

Reply:
Shell didn't parse anything. You are parsing in the `main` function. You are assigning meaning to the inputs and validating them.

Reply:
main() and parseArgs()  just move input arguments from the array to the output list https://github.com/aws-greengrass/aws-greengrass-client-device-auth/blob/uat-dev/uat/custom-components/client-java-sdk/src/main/java/com/aws/greengrass/testing/mqtt5/client/Main.java#L80

to `void main(String[] args)` IP addresses came as args[2], args[3], args[4] separately.
Double quote is magic here.